### PR TITLE
bpo-46523: fix tests rerun when `setUp[Class|Module]` fails

### PR DESCRIPTION
**Pull request by sobolevn**
## Problem

There were two problems:
1. Tests were trying to rerun `setUpClass` as a test name. It is visible here: `Re-running test_typing in verbose mode (matching: setUpClass)`
2. Rerun was not considered an error even if there were errors.

I've fixed both problems.

### setUpModule

```
» ./python.exe -m test -w -v test_typing
== CPython 3.11.0a4+ (heads/main-dirty:ef3ef6fa43, Jan 20 2022, 20:48:25) [Clang 11.0.0 (clang-1100.0.33.16)]
== macOS-10.14.6-x86_64-i386-64bit little-endian
== cwd: /Users/sobolev/Desktop/cpython/build/test_python_6745æ
== CPU count: 4
== encodings: locale=UTF-8, FS=utf-8
0:00:00 load avg: 9.85 Run tests sequentially
0:00:00 load avg: 9.85 [1/1] test_typing
setUpModule (test.test_typing) ... ERROR

======================================================================
ERROR: setUpModule (test.test_typing)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sobolev/Desktop/cpython/Lib/test/test_typing.py", line 43, in setUpModule
    raise TypeError('mo')
    ^^^^^^^^^^^^^^^^^^^^^
TypeError: mo

----------------------------------------------------------------------
Ran 0 tests in 0.004s

FAILED (errors=1)
test test_typing failed
test_typing failed (1 error)

== Tests result: FAILURE ==

1 test failed:
    test_typing
0:00:00 load avg: 9.85
0:00:00 load avg: 9.85 Re-running failed tests in verbose mode
0:00:00 load avg: 9.85 Re-running test_typing in verbose mode (matching: test_typing)
setUpModule (test.test_typing) ... ERROR

======================================================================
ERROR: setUpModule (test.test_typing)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sobolev/Desktop/cpython/Lib/test/test_typing.py", line 43, in setUpModule
    raise TypeError('mo')
    ^^^^^^^^^^^^^^^^^^^^^
TypeError: mo

----------------------------------------------------------------------
Ran 0 tests in 0.003s

FAILED (errors=1)
test test_typing failed
1 test failed again:
    test_typing

== Tests result: FAILURE then FAILURE ==

1 test failed:
    test_typing

1 re-run test:
    test_typing

Total duration: 378 ms
Tests result: FAILURE then FAILURE
```

### setUpClass

https://gist.github.com/sobolevn/adcaf0d470226b8644f70ef81178e028

### setUp

https://gist.github.com/sobolevn/acf04e0f7ea785e1fa4c45a94f545743

https://bugs.python.org/issue46523

### CODE REVIEW

File: Lib/test/test_typing.py [hunk 1]
```diff
@@ -4001,6 +4001,7 @@ def cleanup(self):
 
     @classmethod
     def setUpClass(cls):
+        raise ValueError
```
Comment (sobolevn) [inline Lib/test/test_typing.py]: This is intentional. Let's see the CI output, it should be red.

Comment (sobolevn) [reply Lib/test/test_typing.py]: It failed as intended: https://github.com/python/cpython/runs/4943128810?check_suite_focus=true#step:5:1483 🎉

**Comment by vstinner** [comment]
@ambv: This change is about the code that you wrote if I understood correctly.

**Comment by MaxwellDupre** [comment]
I cant reproduce in 3.12.0a0 latest.
Also, bit confused your command line uses .exe but output is with Linux type directories.
The use case is not clear to me.

**Comment by sobolevn** [comment]
> I cant reproduce in 3.12.0a0 latest.

I haven't tried it on 3.12, only 3.11.x

> Also, bit confused your command line uses .exe but output is with Linux type directories.

I am on macos, since we have `Python/` directory and macos does not recognise `./python` executable and `Python/` directory at the same time, we have to use `.exe` extension - just like Windows does.

> The use case is not clear to me.

There was a problem with how tests were re-run. See https://github.com/python/cpython/issues/90681#issuecomment-1093943320 for more details.

File: Lib/test/libregrtest/main.py [hunk 2]
```diff
@@ -28,6 +28,14 @@
 # Must be smaller than buildbot "1200 seconds without output" limit.
 EXIT_TIMEOUT = 120.0
 
+# bpo-46523: When rerunning tests, we might need to rerun the whole
+# class or module suite if some its life-cycle hooks fail.
+_TEST_LIFECYCLE_HOOKS = frozenset((
```
Comment (sobolevn) [inline Lib/test/libregrtest/main.py:35]: I think this also needs to include `async` methods as well.

**Comment by sobolevn** [comment]
I can merge `main` later tomorrow to get this back into shape if someone wants to get this fixed :)

**Comment by vstinner** [comment]
If you want me to review, I would feel more comfortable with a new test added to test_regrtest. This code is tricky and doesn't seem to be carefully tested.

**Comment by sobolevn** [comment]
Sure thing, will do! 👍 
Happy to see that you are back!

**Comment by sobolevn** [comment]
@vstinner done! 

I've added 8 new test cases:
- `setUpClass` and `tearDownClass` (were affected, now fixed)
- `setUpModule` and `tearDownModule` (were affected, now fixed)
- `setUp` and `tearDown` (were not affected, were not tested, now are tested)
- `asyncSetUp` and `asyncTearDown` (were not affected, were not tested, now are tested)

File: Lib/test/libregrtest/main.py (hunk 2 above)
Comment (sobolevn) [reply Lib/test/libregrtest/main.py:35]: Nope, instead I remove `setUp` and `tearDown`.

**Comment by ambv** [comment]
3.10 is out of scope already. 3.11 I'll merge myself.

**Comment by sobolevn** [comment]
Thanks everyone! 🎉 
I am always happy when some old PR gets merged 😊

**Comment by sobolevn** [comment]
See https://github.com/python/cpython/pull/103170 for this buildbot failure.

**Pull request merged**